git-worktree/src/lib.rs

```diff
@@ -1 +1,126 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_object::bstr::ByteSlice;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::time::Duration;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) {
+            from()
+            display("Could not convert path to UTF8: {}", err)
+        }
+        TimeError(err: std::time::SystemTimeError) {
+            from()
+            display("Could not read file time in proper format: {}", err)
+        }
+        ToU32Error(err: std::num::TryFromIntError) {
+            from()
+            display("Could not convert seconds to u32: {}", err)
+        }
+        IoError(err: std::io::Error) {
+            from()
+            display("IO Error: {}", err)
+        }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut git_index::State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<git_object::BlobRef<'a>>,
+{
+    let path = path.as_ref();
+    let mut buf = Vec::new();
+    let mut entry_time = Vec::new(); // Entries whose timestamps have to be updated
+    for (i, entry) in state.entries().iter().enumerate() {
+        if entry.flags.contains(git_index::entry::Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let dest = path.join(entry.path(state).to_path()?);
+        create_dir_all(dest.parent().expect("path is empty"))?;
+        match entry.mode {
+            git_index::entry::Mode::FILE | git_index::entry::Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                std::fs::write(&dest, obj.data)?;
+                if entry.mode == git_index::entry::Mode::FILE_EXECUTABLE {
+                    #[cfg(unix)]
+                    fs::set_permissions(&dest, fs::Permissions::from_mode(0o777)).unwrap();
```

Comment:
There should never be a panic if something fails, but instead of avoiding `unwrap` here, I think file creation/opening should take care of the execute bit like [git does](https://github.com/git/git/blob/main/entry.c#L81:L82). This can be done with [`custom_flags()`](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.custom_flags).

Reply:
Is this something you tried?

Reply:
I tried implementing you suggestion and was able to but it made that part of the code more complicated.
So, I decided to set the permission bit after writing. Is this fine ?

Reply:
Yes, please, use `custom_flags()` - let's not use more IO operations than git for critical functionality like this.

Reply:
Fixed: https://github.com/Byron/gitoxide/commit/e838eaa5721d8b1b13155aa81234c9c44d9b15fe
Reduces 2 io calls (metadata and permissions)